src/frontend/src/optimizer/plan_node/stream_sink.rs

```diff
@@ -315,6 +318,56 @@ impl StreamSink {
         let (pk, _) = derive_pk(input.clone(), user_order_by, &columns);
         let mut downstream_pk =
             Self::parse_downstream_pk(&columns, properties.get(DOWNSTREAM_PK_KEY))?;
+
+        {
+            // check for key encode TEXT
+            // if key encode is TEXT, we require the key to be a single column
+            // and the column should be of these types: varchar, bool, small int, int, big int
+            if let Some(format_desc) = &format_desc
+                && let Some(key_encode) = &format_desc.key_encode
```

Comment:
Encoder-specific check in `EncoderBuild::build()` and `::new()`. This file can be untouched.